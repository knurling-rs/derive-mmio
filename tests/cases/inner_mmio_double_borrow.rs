#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct UartBank {
    // this is read-write by default
    data: u32,
    status: u32,
}

#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct Uart {
    // you can be explicit if you like
    control: u32,
    #[mmio(inner)]
    bank_0: UartBank,
    #[mmio(inner)]
    bank_1: UartBank,
}

fn main() {
    let mut uart = Uart {
        control: 0xC,
        bank_0: UartBank {
            data: 0x1,
            status: 0x2,
        },
        bank_1: UartBank {
            data: 0x3,
            status: 0x4,
        },
    };

    // Safety: We're pointing at a real object
    let mut mmio_uart = unsafe { Uart::new_mmio(core::ptr::addr_of_mut!(uart)) };
    let mut bank0 = mmio_uart.bank_0();
    let _ = mmio_uart.bank_1().read_data();
    let _ = bank0.read_data();
}
