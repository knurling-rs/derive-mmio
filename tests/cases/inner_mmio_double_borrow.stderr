error[E0499]: cannot borrow `mmio_uart` as mutable more than once at a time
  --> tests/cases/inner_mmio_double_borrow.rs:36:13
   |
35 |     let mut bank0 = mmio_uart.bank_0();
   |                     --------- first mutable borrow occurs here
36 |     let _ = mmio_uart.bank_1().read_data();
   |             ^^^^^^^^^ second mutable borrow occurs here
37 |     let _ = bank0.read_data();
   |             ----- first borrow later used here
